src/Internal/DependencyManagement/AbstractServiceProvider.php

```diff
@@ -0,0 +1,119 @@
+<?php
+/**
+ * AbstractServiceProvider class file.
+ *
+ * @package Automattic\WooCommerce\Internal\DependencyManagement
+ */
+
+namespace Automattic\WooCommerce\Internal\DependencyManagement;
+
+use League\Container\Argument\RawArgument;
+use League\Container\Definition\DefinitionInterface;
+use League\Container\Definition\Definition;
+
+/**
+ * Base class for the service providers used to register classes in the container.
+ *
+ * See the documentation of the original class this one is based on (https://container.thephpleague.com/3.x/service-providers)
+ * for basic usage details. What this class adds is:
+ *
+ * - The `add_with_auto_arguments` method that allows to register classes without having to specify the constructor arguments.
+ * - The `share_with_auto_arguments` method, sibling of the above.
+ * - Convenience `add` and `share` methods that are just proxies for the same methods in `$this->getContainer()`.
+ */
+abstract class AbstractServiceProvider extends \League\Container\ServiceProvider\AbstractServiceProvider {
+
+	/**
+	 * Register a class in the container and use reflection to guess the constructor arguments.
+	 *
+	 * WARNING: this method uses reflection, so please have performance in mind when using it.
+	 *
+	 * @param string $class_name Class name to register.
+	 * @param mixed  $concrete   The concrete to register. Can be a shared instance, a factory callback, or a class name.
+	 * @param bool   $shared Whether to register the class as shared (`get` always returns the same instance) or not.
+	 *
+	 * @return DefinitionInterface The generated container definition.
+	 *
+	 * @throws \Exception Error when reflecting the class, or class constructor is not public, or an argument has no valid type hint.
+	 */
+	protected function add_with_auto_arguments( string $class_name, $concrete = null, bool $shared = false ) : DefinitionInterface {
```

Comment:
It's worth noting that classes can't be resolved without registering them in the container. The standard way to register them involves manually defining all of the dependencies that need to be resolved. This method allows us to define what the dependencies of the class are via reflection as opposed to manually. While in general it's advisable to just define the arguments ourselves this can be useful during experimentation.

A closure concrete on the other hand still requires you to manually define the dependencies.

Reply:
Ah ok. resolving this conversation. we can manually define dependencies using the `add` method if we need to anyway.